fvm/src/blockstore/buffered.rs

```diff
@@ -43,15 +42,10 @@ where
 {
     /// Flushes the buffered cache based on the root node.
     /// This will recursively traverse the cache and write all data connected by links to this
-    /// root Cid. Calling flush will not reset the write buffer.
+    /// root Cid, moving the reachable blocks from the write buffer to the backing store.
     fn flush(&self, root: &Cid) -> Result<()> {
-        let mut buffer = Vec::new();
-        let s = self.write.borrow();
-        copy_rec(&s, *root, &mut buffer)?;
-
-        self.base.put_many_keyed(buffer)?;
-
-        Ok(())
+        self.base
+            .put_many_keyed(take_reachable(&mut self.write.borrow_mut(), root)?)
```

Comment:
I changed this to "take" blocks so we wouldn't end up writing them multiple times. The alternative would have been to keep some form of "written" set.

For some history, we used to clear the _entire_ buffer on flush. However, _that_ ended up causing some issues if we, e.g., wanted to flush some intermediate state. Now we take a middle-path of deleting the written blocks but leaving everything else.

Reply:
Will this then end up saving disk space?